mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -0,0 +1,571 @@
+    // residual evaluation for Iceberg Generics in InputFormat
+    InMemoryDataModel dataModel = conf.getEnum(IN_MEMORY_DATA_MODEL, InMemoryDataModel.DEFAULT);
+    if (dataModel == InMemoryDataModel.DEFAULT || !platformAppliesFilter) {
+      task.files().forEach(fileScanTask -> {
+        Expression residual = fileScanTask.residual();
+        if (residual != null && !residual.equals(Expressions.alwaysTrue())) {
+          throw new RuntimeException(
+              String.format(
+                  "Filter expression %s is not completely satisfied. Additional rows " +
+                      "can be returned not satisfied by the filter expression", residual));
+        }
+      });
+    }
+  }
+
+  @Override
+  public RecordReader<Void, T> createRecordReader(InputSplit split, TaskAttemptContext context) {
+    return new IcebergRecordReader<>();
+  }
+
+  private static final class IcebergRecordReader<T> extends RecordReader<Void, T> {
+    private TaskAttemptContext context;
+    private Iterator<FileScanTask> tasks;
+    private Iterator<T> currentIterator;
+    private T currentRow;
+    private Schema expectedSchema;
+    private Schema tableSchema;
+    private InMemoryDataModel inMemoryDataModel;
+    private Closeable currentCloseable;
+    private boolean reuseContainers;
+    private boolean caseSensitive;
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext newContext) {
+      Configuration conf = newContext.getConfiguration();
+      // For now IcebergInputFormat does its own split planning and does not
+      // accept FileSplit instances
+      CombinedScanTask task = ((IcebergSplit) split).task;
+      this.context = newContext;
+      this.tasks = task.files().iterator();
+      this.tableSchema = SchemaParser.fromJson(conf.get(TABLE_SCHEMA));
+      String readSchemaStr = conf.get(READ_SCHEMA);
+      if (readSchemaStr != null) {
+        this.expectedSchema = SchemaParser.fromJson(readSchemaStr);
+      }
+      this.reuseContainers = conf.getBoolean(REUSE_CONTAINERS, false);
+      this.caseSensitive = conf.getBoolean(CASE_SENSITIVE, true);
+      this.inMemoryDataModel = conf.getEnum(IN_MEMORY_DATA_MODEL, InMemoryDataModel.DEFAULT);
+      this.currentIterator = open(tasks.next());
+    }
+
+    @Override
+    public boolean nextKeyValue() throws IOException {
+      while (true) {
+        if (currentIterator.hasNext()) {
+          currentRow = currentIterator.next();
+          return true;
+        } else if (tasks.hasNext()) {
+          currentCloseable.close();
+          currentIterator = open(tasks.next());
+        } else {
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public Void getCurrentKey() {
+      return null;
+    }
+
+    @Override
+    public T getCurrentValue() {
+      return currentRow;
+    }
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
```

Comment:
Maybe leave a TODO item to improve this.

Reply:
@rdblue , any thoughts on what we need to improve here?

Reply:
Just that we could give a more accurate progress based on records read from the file. I don't think that `context.getProgress` has enough information to give an accurate progress value.

This isn't that easy, since we don't know how much of the input split has been processed and we are pushing filters into Parquet and ORC. But we do know when a file is opened and could count the number of rows returned, so we can estimate. And we could also add a row count to the readers so that we can get an accurate count of rows that have been either returned or filtered out.

Reply:
Added